crates/ruff_server/src/session/settings.rs

```diff
@@ -1,25 +1,284 @@
+/// which inherits the schema of [`UserSettings`] and adds extra fields
+/// to describe the workspace it applies to.
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct WorkspaceSettings {
+    #[serde(flatten)]
+    user_settings: UserSettings,
+    workspace: Url,
+}
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(transparent)]
+#[repr(transparent)]
+struct FixAll(bool);
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(transparent)]
+#[repr(transparent)]
+struct OrganizeImports(bool);
+
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct Lint {
+    enable: Option<LintEnable>,
+    run: Option<RunWhen>,
+}
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(transparent)]
+#[repr(transparent)]
+struct LintEnable(bool);
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+enum RunWhen {
+    OnType,
+    OnSave,
+}
+
+#[derive(Debug, Default, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct CodeAction {
+    disable_rule_comment: Option<CodeActionSettings>,
+    fix_violation: Option<CodeActionSettings>,
+}
+
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct CodeActionSettings {
+    enable: Option<CodeActionEnable>,
+}
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(transparent)]
+#[repr(transparent)]
+struct CodeActionEnable(bool);
+
+#[derive(Clone, Copy, Debug, Default, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "lowercase")]
+enum LogLevel {
+    #[default]
+    Error,
+    Warn,
+    Info,
+    Debug,
+}
+
+/// This is the exact schema for initialization options sent in by the client
+/// during initialization.
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(untagged)]
+enum InitializationOptions {
+    #[serde(rename_all = "camelCase")]
+    HasWorkspaces {
+        global_settings: UserSettings,
+        #[serde(rename = "settings")]
+        workspace_settings: Vec<WorkspaceSettings>,
+    },
+    GlobalOnly {
+        settings: Option<UserSettings>,
+    },
+}
+
+impl SettingsController {
+    /// Initializes the controller from the serialized initialization options.
+    /// This fails if `options` are not valid initialization options.
+    pub(crate) fn from_value(options: serde_json::Value) -> crate::Result<Self> {
+        let options = serde_json::from_value(options).map_err(|err| {
+            anyhow::anyhow!("Failed to deserialize initialization options: {err}")
+        })?;
+        Ok(Self::from_init_options(options))
+    }
```

Comment:
This is unrelated to the PR but I want to understand how are these log messages being sent to the client. Currently, we use `tracing` which I think just dumps the messsages to stderr which is what the client is picking up?

It's not required to be done in this PR but it would be useful to move to providing useful messages to the user using either of the following:
1. [Show message notification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_showMessage)
2. [Show message request](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_showMessageRequest)
3. [Log message](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_logMessage)

For `ruff-lsp`, we use a combination of (2) and (3) depending on user config and the log level.

Reply:
Right now, we using the `tracing` library for logging, which writes out to `stderr`. This does show up in the `Output` window in the editor, though it's always tagged as `[info]` and it would be nice to have log messages tagged properly. Switching over to use `window/logMessage` has actually been a side project that I've made some progress on this week. I hope to open some PRs in the near future.